File: leetcode/array/leetcode_525.java
package leetcode.array;

import java.util.HashMap;
import java.util.Map;

public class leetcode_525 {
    public static void main(String[] args) {
        leetcode_525 lc = new leetcode_525();
        int[] nums = {0, 1, 0};
        System.out.println(lc.findMaxLength(nums));
    }

    public int findMaxLength(int[] nums) {
        Map<Integer, Integer> map = new HashMap<>();
        map.put(0, -1);

        int sum = 0, max = 0;

        for (int i = 0; i < nums.length; i++) {
            sum += (nums[i] == 0) ? -1 : 1;
            if (!map.containsKey(sum)) map.put(sum, i);
            max = Math.max(max, i - map.get(sum));
        }

        return max;
    }
}

File: array/LeetCode_525.kt
package array

class LeetCode_525 {

    fun findMaxLength(nums: IntArray): Int {
        val lengthMap = hashMapOf<Int, Int>()
        lengthMap[0] = -1

        var sum = 0
        var max = 0

        for (i in nums.indices) {
            sum += if (nums[i] == 0) -1 else 1

            if (!lengthMap.containsKey(sum)) {
                lengthMap[sum] = i
            }

            max = maxOf(max, i - lengthMap[sum]!!)

        }

        return max
    }
}

fun main() {
    // Input
    val nums = intArrayOf(0, 1)
    // Output
    LeetCode_525().findMaxLength(nums).run {
        println(this)
        require(2 == this)
    }
}
